### PR TITLE
Parquet performance improvements for write

### PR DESCRIPTION
**Pull request by frankgh**
Improve performance during Parquet write. We no longer split files that
go over 128MB into multiple files, we also use new parquet libraries.
Finally, we refactor the parquet write code to use non-deprecated API
calls.

A new parameter `ENABLE_DICTIONARY` is also added. Parquet dictionaries
are enabled by default (no behavior change from previous versions) but
the dictionary can be disabled by providing `ENABLE_DICTIONARY=false`

### CODE REVIEW

File: server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java [hunk 1]
```diff
@@ -0,0 +1,942 @@
...
+
+        MessageType schema = validateFooter(expectedFile);
+
+        ParquetReader<Group> fileReader = ParquetReader.builder(new GroupReadSupport(), expectedFile)
+                .withConf(configuration)
+                .build();
+
+        // Physical type is binary, logical type is Date
+        assertTrue(schema.getType(0).getLogicalTypeAnnotation() instanceof LogicalTypeAnnotation.StringLogicalTypeAnnotation);
+        assertEquals("2020-08-01", fileReader.read().getString(0, 0));
```
Comment (denalex) [inline server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:322]: I am not sure we are correctly storing DATE as STRING: 
https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#date

File: server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java [hunk 2]
```diff
@@ -0,0 +1,942 @@
...
+        Path expectedFile = new Path(HcfsType.LOCALFILE.getUriForWrite(configuration, context, true) + ".snappy.parquet");
+        assertTrue(expectedFile.getFileSystem(configuration).exists(expectedFile));
+
+        MessageType schema = validateFooter(expectedFile);
+
+        ParquetReader<Group> fileReader = ParquetReader.builder(new GroupReadSupport(), expectedFile)
+                .withConf(configuration)
+                .build();
+
+        // Physical type is binary, logical type is Date
```
Comment (denalex) [inline server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:368]: copy & paste issue ?

File: server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java [hunk 3]
```diff
@@ -0,0 +1,942 @@
...
+
+        context.setDataSource(path);
+        context.setTransactionId("XID-XYZ-123459");
+
+        accessor.initialize(context);
+        resolver.initialize(context);
+
+        assertTrue(accessor.openForWrite());
+
+        // write parquet file with DATE from 2020-08-01 to 2020-08-10
```
Comment (denalex) [inline server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:349]: copy & paste ?

File: server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java [hunk 4]
```diff
@@ -0,0 +1,942 @@
...
+        assertTrue(expectedFile.getFileSystem(configuration).exists(expectedFile));
+
+        MessageType schema = validateFooter(expectedFile);
+
+        ParquetReader<Group> fileReader = ParquetReader.builder(new GroupReadSupport(), expectedFile)
+                .withConf(configuration)
+                .build();
+
+        // Physical type is INT
+        assertTrue(schema.getType(0).getLogicalTypeAnnotation() instanceof LogicalTypeAnnotation.IntLogicalTypeAnnotation);
```
Comment (denalex) [inline server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:615]: why there's an annotation if the physical type is already int ? I assume int32 ?

File: server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java [hunk 5]
```diff
@@ -0,0 +1,942 @@
...
+        assertTrue(expectedFile.getFileSystem(configuration).exists(expectedFile));
+
+        MessageType schema = validateFooter(expectedFile);
+
+        ParquetReader<Group> fileReader = ParquetReader.builder(new GroupReadSupport(), expectedFile)
+                .withConf(configuration)
+                .build();
+
+        // Physical type is BINARY
+        assertTrue(schema.getType(0).getLogicalTypeAnnotation() instanceof LogicalTypeAnnotation.StringLogicalTypeAnnotation);
```
Comment (denalex) [inline server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:712]: can we actually assert the actual type coming back for each of these methods ?

File: server/pxf-hdfs/src/main/java/org/greenplum/pxf/plugins/hdfs/parquet/ParquetTypeConverter.java [hunk 6]
```diff
@@ -32,17 +32,15 @@
...
                 return DataType.BYTEA;
-            }
-            switch (originalType) {
-                case DATE:
-                    return DataType.DATE;
-                case TIMESTAMP_MILLIS:
-                    return DataType.TIMESTAMP;
-                default:
-                    return DataType.TEXT;
+            } else if (originalType instanceof LogicalTypeAnnotation.DateLogicalTypeAnnotation) {
```
[the remark this answers is not in the document]
Comment (frankgh) [reply server/pxf-hdfs/src/main/java/org/greenplum/pxf/plugins/hdfs/parquet/ParquetTypeConverter.java:38]: `LogicalTypeToken.getType()` is package-private unfortunately, so we have to use `instanceof`. For example:

```
    @Override
    LogicalTypeToken getType() {
      return LogicalTypeToken.DATE;
    }
```

File: server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java (hunk 1 above)
Comment (frankgh) [reply server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:322]: I can investigate and fix in a follow up PR

File: server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java (hunk 3 above)
Comment (frankgh) [reply server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:349]: good catch

File: server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java (hunk 4 above)
Comment (frankgh) [reply server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:615]: Physical type is int32, but you can represent int32, int16 and int8 with the physical int32 type

File: server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java (hunk 5 above)
Comment (frankgh) [reply server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:712]: I'm not sure what you mean

File: server/pxf-hdfs/src/main/java/org/greenplum/pxf/plugins/hdfs/ParquetFileAccessor.java [hunk 7]
```diff
@@ -490,33 +493,29 @@ private MessageType generateParquetSchema(List<ColumnDescriptor> columns) {
...
+                            .length(PRECISION_TO_BYTE_COUNT[precision - 1])
+                            .as(LogicalTypeAnnotation.DecimalLogicalTypeAnnotation.decimalType(scale, precision));
                     break;
                 case TIMESTAMP:
                 case TIMESTAMP_WITH_TIME_ZONE:
-                    typeName = PrimitiveTypeName.INT96;
+                    builder = Types.optional(PrimitiveTypeName.INT96);
                     break;
                 case DATE:
                 case TIME:
```
[the remark this answers is not in the document]
Comment (frankgh) [reply server/pxf-hdfs/src/main/java/org/greenplum/pxf/plugins/hdfs/ParquetFileAccessor.java:506]: We'll need to double check dates / times. I will investigate and address in a follow up PR. Timestamp with timezone is converted to UTC here `ParquetTypeConverter.getBinaryFromTimestampWithTimeZone`

File: server/pxf-hdfs/src/main/java/org/greenplum/pxf/plugins/hdfs/ParquetFileAccessor.java [hunk 8]
```diff
@@ -490,33 +493,29 @@ private MessageType generateParquetSchema(List<ColumnDescriptor> columns) {
...
-                    dmt = new DecimalMetadata(precision, scale);
+                    builder = Types
+                            .optional(PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY)
+                            .length(PRECISION_TO_BYTE_COUNT[precision - 1])
+                            .as(LogicalTypeAnnotation.DecimalLogicalTypeAnnotation.decimalType(scale, precision));
                     break;
                 case TIMESTAMP:
                 case TIMESTAMP_WITH_TIME_ZONE:
-                    typeName = PrimitiveTypeName.INT96;
+                    builder = Types.optional(PrimitiveTypeName.INT96);
```
[the remark this answers is not in the document]
Comment (frankgh) [reply server/pxf-hdfs/src/main/java/org/greenplum/pxf/plugins/hdfs/ParquetFileAccessor.java:506]: we have a story called `INT96 deprecated in Parquet 2.5.0`

File: server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java (hunk 1 above)
Comment (frankgh) [reply server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:324]: I just verified that PXF is storing date as String whereas Hive stores date as int32 with logical type DATE.

Comment (frankgh) [reply server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:324]: **Hive**
```
 gpadmin@mdw  ~/parquet  java -jar parquet-tools-1.12.0-SNAPSHOT.jar schema /tmp/000000_0
message hive_schema {
  optional int32 dt (DATE);
}
```

**PXF**
```
 gpadmin@mdw  ~/parquet  java -jar parquet-tools-1.12.0-SNAPSHOT.jar schema /tmp/1597879570-0000000014_2.0.snappy.parquet
message hive_schema {
  optional binary dt (STRING);
}
```

File: server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java [hunk 9]
```diff
@@ -0,0 +1,973 @@
...
+        assertTrue(expectedFile.getFileSystem(configuration).exists(expectedFile));
+
+        MessageType schema = validateFooter(expectedFile);
+
+        ParquetReader<Group> fileReader = ParquetReader.builder(new GroupReadSupport(), expectedFile)
+                .withConf(configuration)
+                .build();
+
+        // There is no logical annotation, the physical type is INT32
+        assertNull(schema.getType(0).getLogicalTypeAnnotation());
```
Comment (denalex) [inline server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java:228]: what I meant here and everywhere else is to add assertion of the actual type:
`assertEquals(<EXPECTED_PHYSICAL_TYPE>, schema.getType(0));`

Comment (frankgh) [reply server/pxf-hdfs/src/test/java/org/greenplum/pxf/plugins/hdfs/ParquetWriteTest.java]: ready!

**Pull request merged**